dash/include/dash/algorithm/Sort.h

```diff
@@ -0,0 +1,1177 @@
+  auto const sz = partitions.size() + 2;
+  // Number of elements less than P
+  std::vector<std::size_t> n_lt(sz, 0);
+  // Number of elements less than equals P
+  std::vector<std::size_t> n_le(sz, 0);
+
+  auto comp_lower = [&sortable_hash](
+                        const ElementType& a, const MappedType& b) {
+    return sortable_hash(const_cast<ElementType&>(a)) < b;
+  };
+
+  auto comp_upper = [&sortable_hash](
+                        const MappedType& b, const ElementType& a) {
+    return b < sortable_hash(const_cast<ElementType&>(a));
+  };
+
+  auto const n_l_elem = std::distance(lbegin, lend);
+
+  if (n_l_elem == 0) return std::make_pair(n_lt, n_le);
+
+  for (auto const& idx : valid_partitions) {
+    auto lb_it = std::lower_bound(lbegin, lend, partitions[idx], comp_lower);
+    auto ub_it = std::upper_bound(lbegin, lend, partitions[idx], comp_upper);
+
+    auto const p_left = p_borders.left_partition[idx];
+    DASH_ASSERT_NE(p_left, dash::team_unit_t{}, "invalid bounding unit");
+
+    n_lt[p_left + 1] = std::distance(lbegin, lb_it);
+    n_le[p_left + 1] = std::distance(lbegin, ub_it);
+  }
+
+  auto const last_valid_border_idx = *std::prev(valid_partitions.cend());
+
+  auto const p_left = p_borders.left_partition[last_valid_border_idx];
+
+  std::fill(
+      std::next(std::begin(n_lt), p_left + 2), std::end(n_lt), n_l_elem);
+  std::fill(
+      std::next(std::begin(n_le), p_left + 2), std::end(n_le), n_l_elem);
+
+  DASH_LOG_TRACE("psort__local_histogram >");
+  return std::make_pair(n_lt, n_le);
+}
+
+template <typename SizeType, typename GlobIter>
+inline void psort__global_histogram(
+    std::vector<SizeType> const& l_nlt,
+    std::vector<SizeType> const& l_nle,
+    std::vector<size_t> const&   valid_partitions,
+    GlobIter                     it_nlt_nle)
+{
+  DASH_LOG_TRACE("< psort__global_histogram ");
+  DASH_ASSERT_EQ(l_nlt.size(), l_nle.size(), "Sizes must match");
+
+  auto const& team = it_nlt_nle.pattern().team();
+  auto const  myid = team.myid();
+
+  auto liter_begin = (it_nlt_nle + (myid.id * NLT_NLE_BLOCK)).local();
+
+  DASH_ASSERT_MSG(liter_begin, "pointer must not be NULL");
+
+  std::fill(liter_begin, liter_begin + NLT_NLE_BLOCK, 0);
+
+  team.barrier();
+
+  auto const last_valid_border = *std::prev(valid_partitions.cend());
+
+  // TODO: implement dash::transform_async
+  // We have to add 2
+  // --> +1 because we start at idx 1
+  // --> +1 to get the last idx
+  for (std::size_t idx = 1; idx < last_valid_border + 2; ++idx) {
+    // we communicate only non-zero values
+    if (l_nlt[idx] == 0 && l_nle[idx] == 0) continue;
+
+    std::array<SizeType, NLT_NLE_BLOCK> vals{{l_nlt[idx], l_nle[idx]}};
+    auto const                          g_idx_nlt = (idx - 1) * NLT_NLE_BLOCK;
+
+    dash::transform(
+        std::begin(vals),  // A
+        std::end(vals),
+        it_nlt_nle + g_idx_nlt,   // B
+        it_nlt_nle + g_idx_nlt,   // B = op(B,A)
+        dash::plus<SizeType>());  // op
```

Comment:
That is probably a scalability problem. Maybe use `dart_accumulate` and `dart_flush` for now?